support 'resample_draws'

paul-buerkner opened the pull request:
This PR implements #54. @MansMeg thank you for agreeing to review this PR!

MansMeg commented on NAMESPACE:
```diff
@@ -89,6 +89,7 @@ S3method(repair_draws,draws_array)
 S3method(repair_draws,draws_df)
 S3method(repair_draws,draws_list)
 S3method(repair_draws,draws_matrix)
+S3method(resample_draws,draws)
```
Do we only want this methods to work with draws objects? Not draws_df, draws_list etc?

paul-buerkner replied:
`draws` is the parent class of all the formats so a `.draws` method will apply to all formats.

MansMeg commented on R/resample_draws.R:
```diff
@@ -0,0 +1,176 @@
+#' @template args-methods-x
+#' @param weights A vector of positive weights of length equal to the
+#' number of draws in `draws` (as obtained by method [`ndraws`]).
+#' Draws with higher weights will be selected more often, usually
```
Is this sentence necessary? 
"Draws with higher weights will be selected more often, usually ... their weights."

paul-buerkner replied:
Probably not. I will remove it.

MansMeg commented on R/resample_draws.R:
```diff
@@ -0,0 +1,176 @@
+#' Draws with higher weights will be selected more often, usually
+#' roughly proportional to their weights. How exactly the weights are handled
+#' depends on the `method` argument. Weights will be internally normalized
+#' if necessary.
```
Will they not always be normalized internally?

paul-buerkner replied:
True. I will correct that.

MansMeg commented on R/resample_draws.R:
```diff
@@ -0,0 +1,176 @@
+#' @param method Name of the resampling method being applied.
+#' Possible choices are `"simple"` for simple random resampling with
+#' replacement, `"simple_no_replace"` for simple random resampling without
+#' replacement, `"stratified"` for stratified resampling with replacement
```
Stratified by what? I guess we - in the long run at least - need to specify what the different approaches are doing?

paul-buerkner replied:
This is a good point. I need to read the paper first :-D For now I just refer to the paper for details and we can change it later on.

MansMeg commented on R/resample_draws.R:
```diff
@@ -0,0 +1,176 @@
+resample_draws.draws <- function(x, weights, method = "stratified",
+                                 ndraws = NULL, ...) {
+  if (!allow_resample_draws(x)) {
+    stop2("'resample_draws' requires formats which can be subsetted ",
```
What is the reason for this?

paul-buerkner replied:
After subsetting draws, the structure will no longer be compatible with a draws_array or draws_list object since, after subsetting, different chains will have different lengths. Do you see a way around this?

paul-buerkner replied:
One option would be to drop the chain dimension (with a message). A similar thing technically happens `draws_df` objects already. I will work on this option.

paul-buerkner replied:
Ok. Should be working now for all formats.

MansMeg commented on R/resample_draws.R:
```diff
@@ -0,0 +1,176 @@
+  weights <- weights / sum(weights)
+  method_fun <- paste0(".resample_", method)
+  method_fun <- get(method_fun, asNamespace("posterior"))
+  sel_draws <- method_fun(
```
maybe `draws_idx` or similar would be a more clear variable name? Now I expect a boolean vector (select/not select)?

paul-buerkner replied:
Good point. I will change that.

paul-buerkner commented:
Thanks @MansMeg for your review! I incorporated all the comments of yours. In particular, resampling should now be supported for all formats, via automatic merging of chains. Would you mind taking another look at the PR? I have also added you as contributor to posterior.

paul-buerkner commented on R/resample_draws.R:
```diff
@@ -0,0 +1,176 @@
+# simple random resampling with replacement
+# @return index vector of length 'ndraws'
+.resample_simple <- function(weights, ndraws, ...) {
+  ndraws <- default_ndraws_resample(ndraws, weights)
```
Done.

paul-buerkner replied:
The reason I was initially handling this different was because `ndraws` default cannot be set for method `simple_no_replace`. I think the new way of handling this exception is now cleaner.

paul-buerkner commented on tests/testthat/test-resample_draws.R:
```diff
@@ -0,0 +1,30 @@
+test_that("resample_draws returns expected format", {
```
Good idea. I have added such tests.

MansMeg commented:
Great! I do not think I have any more suggestions now. Good work @paul-buerkner!

paul-buerkner commented:
Thank you! I really appreciate your comments!
